File: src/types/sale.ts
export interface MedicinePosDto {
  id: number;
  name: string;
  generic_name: string | null;
  unit: string;
  retail_price: number;
  current_stock: number;
  reorder_level: number;
  shelf_location: string | null;
}

export interface ConfirmSaleItemDto {
  medicine_id: number;
  quantity: number;
  item_discount: number;
}

export interface ConfirmSaleDto {
  items: ConfirmSaleItemDto[];
  bill_discount: number;
  tax_enabled: boolean;
  payment_method: string;
  customer_name: string | null;
  customer_phone?: string | null;
  due_date?: string | null;
}

export interface SaleReceiptDto {
  sale_id: number;
  subtotal: number;
  bill_discount: number;
  tax_rate: number;
  tax_amount: number;
  total: number;
  payment_method: string;
  customer_name: string | null;
  item_count: number;
  items: SaleReceiptItemDto[];
}

export interface SaleReceiptItemDto {
  medicine_name: string;
  quantity: number;
  unit_price: number;
  item_discount: number;
  line_total: number;
}

export interface SaleListDto {
  id: number;
  total: number;
  payment_method: string;
  customer_name: string | null;
  item_count: number;
  created_at: string;
  total_returned_qty: number;
  total_refund_amount: number;
}

export interface SaleDetailDto {
  sale: {
    id: number;
    user_id: number;
    subtotal: number;
    bill_discount: number;
    tax_rate: number;
    tax_amount: number;
    total: number;
    payment_method: string;
    customer_name: string | null;
    created_at: string;
  };
  items: SaleDetailItemDto[];
}

export interface SaleDetailItemDto {
  id: number;
  medicine_name: string;
  batch_id: number;
  quantity: number;
  unit_price: number;
  purchase_cost: number;
  item_discount: number;
  line_total: number;
}

export interface OwnerDashboardDto {
  today_sales: number;
  today_profit: number;
  month_sales: number;
  low_stock_count: number;
  expiry_warning_count: number;
  expiry_critical_count: number;
  top_sellers: TopSellerDto[];
}

export interface PharmacistDashboardDto {
  today_sales: number;
  low_stock_count: number;
  expiry_warning_count: number;
  expiry_critical_count: number;
}

export interface TopSellerDto {
  medicine_id: number;
  medicine_name: string;
  total_qty: number;
}

File: src-tauri/src/models/sale.rs
use serde::{Deserialize, Serialize};

/// POS-specific medicine search result — excludes purchase_price always (per D-34/POS-01)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MedicinePosDto {
    pub id: i64,
    pub name: String,
    pub generic_name: Option<String>,
    pub unit: String,
    pub retail_price: f64,
    pub current_stock: i64,
    pub reorder_level: i64,
    pub shelf_location: Option<String>,
}

/// Input item from frontend (what user adds to cart)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmSaleItemDto {
    pub medicine_id: i64,
    pub quantity: i64,
    pub item_discount: f64, // currency amount, capped at unit_price * qty
}

/// Frontend payload for confirming a sale
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfirmSaleDto {
    pub items: Vec<ConfirmSaleItemDto>,
    pub bill_discount: f64, // currency amount, capped at subtotal
    pub tax_enabled: bool,
    pub payment_method: String, // "Cash" | "Card" | "Credit"
    pub customer_name: Option<String>,
    #[serde(default)]
    pub customer_phone: Option<String>,
    #[serde(default)]
    pub due_date: Option<String>,
}

/// Full sale header (DB row)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Sale {
    pub id: i64,
    pub user_id: i64,
    pub subtotal: f64,
    pub bill_discount: f64,
    pub tax_rate: f64,
    pub tax_amount: f64,
    pub total: f64,
    pub payment_method: String,
    pub customer_name: Option<String>,
    pub created_at: String,
}

/// Sale line item (DB row)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleItem {
    pub id: i64,
    pub sale_id: i64,
    pub medicine_id: i64,
    pub batch_id: i64,
    pub quantity: i64,
    pub unit_price: f64,
    pub purchase_cost: f64,
    pub item_discount: f64,
    pub line_total: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleListDto {
    pub id: i64,
    pub total: f64,
    pub payment_method: String,
    pub customer_name: Option<String>,
    pub item_count: i64,
    pub created_at: String,
    pub total_returned_qty: i64,
    pub total_refund_amount: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleDetailDto {
    pub sale: Sale,
    pub items: Vec<SaleDetailItemDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleDetailItemDto {
    pub id: i64,
    pub medicine_name: String,
    pub batch_id: i64,
    pub quantity: i64,
    pub unit_price: f64,
    pub purchase_cost: f64,
    pub item_discount: f64,
    pub line_total: f64,
}

/// DTO returned to frontend after successful sale
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleReceiptDto {
    pub sale_id: i64,
    pub subtotal: f64,
    pub bill_discount: f64,
    pub tax_rate: f64,
    pub tax_amount: f64,
    pub total: f64,
    pub payment_method: String,
    pub customer_name: Option<String>,
    pub item_count: i64,
    pub items: Vec<SaleReceiptItemDto>,
}

/// Individual line item in receipt
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SaleReceiptItemDto {
    pub medicine_name: String,
    pub quantity: i64,
    pub unit_price: f64,
    pub item_discount: f64,
    pub line_total: f64,
}

/// Owner dashboard DTO — includes profit data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OwnerDashboardDto {
    pub today_sales: f64,
    pub today_profit: f64,
    pub month_sales: f64,
    pub low_stock_count: i64,
    pub expiry_warning_count: i64,
    pub expiry_critical_count: i64,
    pub top_sellers: Vec<TopSellerDto>,
}

/// Pharmacist dashboard DTO — NO profit data (D-42)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PharmacistDashboardDto {
    pub today_sales: f64,
    pub low_stock_count: i64,
    pub expiry_warning_count: i64,
    pub expiry_critical_count: i64,
}

/// Top-selling medicine this week
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopSellerDto {
    pub medicine_id: i64,
    pub medicine_name: String,
    pub total_qty: i64,
}

/// Internal allocation record (used during FIFO processing, not IPC)
#[derive(Debug, Clone)]
pub struct SaleItemAllocation {
    pub medicine_id: i64,
    pub batch_id: i64,
    pub quantity: i64,
    pub unit_price: f64,
    pub purchase_cost: f64,
    pub item_discount: f64,
    pub line_total: f64,
}
